lib/friendly_shipping/services/ups/parse_xml_response.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module FriendlyShipping
+  module Services
+    class Ups
+      class ParseXMLResponse
+        extend Dry::Monads::Result::Mixin
+
+        class <<self
+          def call(response_body, expected_root_tag)
+            xml = Nokogiri.XML(response_body)
+            if xml.root.nil? || xml.root.name != expected_root_tag
+              Failure('Invalid document')
+            end
+            if request_successful?(xml)
+              Success(xml)
+            else
+              Failure(error_message(xml))
+            end
+          rescue Nokogiri::XML::SyntaxError => e
+            Failure(e)
+          end
+
+          private
+
+          def request_successful?(xml)
+            xml.root.at('Response/ResponseStatusCode').text == '1'
+          end
+
+          def error_message(xml)
+            status = xml.root.at_xpath('Response/ResponseStatusDescription').try(:text)
+            desc = xml.root.at_xpath('Response/Error/ErrorDescription').try(:text)
+            [status, desc].select(&:present?).join(": ").presence || 'UPS could not process the request.'
```

Comment:
`.presence` and `present` are a ActiveSupport features. Do we actually need to check for `empty` as well? Aren't empty XML nodes `nil`? I know that not using ActiveSupport here would lead to more verbose code, but currently `friendly_shipping` does not depend on `activesupport`. Maybe you want to be more explicit on that then?